core/controllers/learner_group.py

```diff
@@ -0,0 +1,500 @@
+                learner_group.invited_student_user_ids),
+            'subtopic_page_ids': learner_group.subtopic_page_ids,
+            'story_ids': learner_group.story_ids
+        })
+
+    @acl_decorators.can_access_learner_groups
+    def delete(self, learner_group_id):
+        """Deletes a learner group."""
+
+        is_valid_request = learner_group_services.is_user_facilitator(
+            self.user_id, learner_group_id
+        )
+
+        if not is_valid_request:
+            raise self.UnauthorizedUserException(
+                'You do not have the rights to delete this learner group '
+                'as you are not its facilitator.')
+
+        learner_group_services.remove_learner_group(learner_group_id)
+
+        self.render_json({
+            'success': True
+        })
+
+
+class LearnerGroupStudentProgressHandler(base.BaseHandler):
+    """Handles operations related to the learner group student's progress."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'learner_group_id': {
+            'schema': {
+                'type': 'basestring',
+                'validators': [{
+                    'id': 'is_regex_matched',
+                    'regex_pattern': constants.LEARNER_GROUP_ID_REGEX
+                }]
+            },
+            'default_value': None
+        }
+    }
+
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {
+            'student_usernames': {
+                'schema': {
+                    'type': 'custom',
+                    'obj_type': 'JsonEncodedInString'
+                }
+            }
+        }
+    }
+
+    @acl_decorators.can_access_learner_groups
+    def get(self, learner_group_id):
+        """Handles GET requests for users progress through learner
+        group syllabus.
+        """
+
+        student_usernames = self.normalized_request.get('student_usernames')
+
+        student_user_ids = user_services.get_multi_user_ids_from_usernames(
+            student_usernames)
+
+        learner_group = learner_group_fetchers.get_learner_group_by_id(
+            learner_group_id)
+
+        if learner_group is None:
+            raise self.InvalidInputException('No such learner group exists.')
+
+        subtopic_page_ids = learner_group.subtopic_page_ids
+        story_ids = learner_group.story_ids
+        topic_ids = (
+            subtopic_page_services.get_topic_ids_from_subtopic_page_ids(
+                subtopic_page_ids
+            )
+        )
+        topics = topic_fetchers.get_topics_by_ids(topic_ids)
+
+        all_skill_ids_lists = [
+            topic.get_all_skill_ids() for topic in topics if topic
+        ]
+        all_skill_ids = list(
+            {
+                skill_id for skill_list in all_skill_ids_lists
+                for skill_id in skill_list
+            }
+        )
+
+        all_students_progress = []
+
+        for user_id in student_user_ids:
+            progress_sharing_permission = (
```

Comment:
I left a comment here but it disappeared for some reason. 

What is the maximum number of students and stories in a group? What is the expected average number? This code is not extremely efficient, but that might be fine if we have some limits on the sizes of the groups. If we don't then we should rewrite it in a way that we only fetch everything once, do the required operations and then save everything at once.

Reply:
We did not have any discussion regarding the maximum number of students and stories in the group, but it is expected to not be a large number because the learner groups are aimed to be for targeted learning and large number of stories in syllabus may disrupt the results.
I personally do not want to revamp this code right now as it is as major chunk of logic and it will take quite some time to modify it and will effect a lot on the project time schedule.

Reply:
Does this for loop make N calls to the data store? If so, we can't check this in. It has to be updated so that it makes a single call to the data store using e.g. get-multi methods, otherwise it will time out in production.

This shouldn't be a new concept...I think I made a similar comment on one of your earlier PRs for this project. The general rule is that no controller should be doing O(N) operations, so please make sure this is not happening anywhere.

Reply:
Modified the progress fetching logic to fetch all users progress data at once.